core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -388,17 +742,30 @@ private SortedRangeSet checkCompatibility(ValueSet other)
         return (SortedRangeSet) other;
     }
 
-    private void checkTypeCompatibility(Marker marker)
-    {
-        if (!getType().equals(marker.getType())) {
-            throw new IllegalStateException(format("Marker of %s does not match SortedRangeSet of %s", marker.getType(), getType()));
-        }
-    }
-
     @Override
     public int hashCode()
     {
-        return Objects.hash(lowIndexedRanges);
+        int hash = lazyHash;
+        if (hash == 0) {
+            hash = Objects.hash(type, Arrays.hashCode(inclusive));
+            for (int position = 0; position < sortedRanges.getPositionCount(); position++) {
+                if (sortedRanges.isNull(position)) {
+                    hash = hash * 31;
+                    continue;
+                }
+                try {
+                    hash = hash * 31 + (int) (long) hashCodeOperator.invokeExact(sortedRanges, position);
+                }
+                catch (Throwable throwable) {
+                    throw handleThrowable(throwable);
+                }
+            }
+            if (hash == 0) {
```

Comment:
Is this necessary? Hashing something for the second time with a chance of 1/4294967296 is not that much of a burden.

Reply:
i agree. this is more like a convention, see https://github.com/trinodb/trino/pull/5885#discussion_r520026604